### PR TITLE
New post: Accelerating PyTorch distributed fine-tuning with Intel technologies

### PR DESCRIPTION
**Pull request by juliensimon**
In this post, you will learn how to accelerate PyTorch training jobs by distributing them on a cluster of Intel Xeon Scalable CPU servers, powered by the Ice Lake architecture and running performance-optimized software libraries. We will build the cluster from scratch using virtual machines, and you should able to easily replicate the demo on your own infrastructure, either in the cloud or on premise.

### CODE REVIEW

File: intel_dist_training.md [hunk 1]
```diff
@@ -0,0 +1,341 @@
+---
```
Comment (julien-c) [inline intel_dist_training.md:1]: this file's name will need to be the url slug that the blog post will live under

File: intel_dist_training.md [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+Graphical Processing Units (GPUs) have long been the _de facto_ choice to train deep learning models. However, the rise of transfer learning is changing the game. Models are now rarely trained from scratch on humungous datasets. Instead, they are frequently fine-tuned on specific (and smaller) datasets, in order to build specialized models that are more accurate than the base model for particular tasks. As these training jobs are much shorter, using a CPU-based cluster can prove to be an interesting option that keeps both training time and cost under control.
+
+### What this post is about
+
+In this post, you will learn how to accelerate [PyTorch](https://pytorch.org) training jobs by distributing them on a cluster of Intel Xeon Scalable CPU servers, powered by the Ice Lake architecture and running performance-optimized software libraries. We will build the cluster from scratch using virtual machines, and you should able to easily replicate the demo on your own infrastructure, either in the cloud or on premise.
```
Comment (jeffboudier) [inline intel_dist_training.md:34]:
```suggestion
In this post, you will learn how to accelerate [PyTorch](https://pytorch.org) training jobs by distributing them on a cluster of Intel Xeon Scalable CPU servers, powered by the Ice Lake architecture and running performance-optimized software libraries. We will build the cluster from scratch using virtual machines, and you should be able to easily replicate the demo on your own infrastructure, either in the cloud or on premise.
```

File: intel_dist_training.md [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+```
+pip install transformers datasets
+yes | conda install scipy scikit-learn
+```
+
+Finally, we clone a fork of the ```transformers```repository containing the example we're going to run.
+
+```
+git clone https://github.com/kding1/transformers.git
```
Comment (hshen14) [inline intel_dist_training.md:181]: maybe add a comment to use your own cloned repo

File: intel_dist_training.md [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+@@ -191,6 +193,17 @@ def main():
+     # or by passing the --help flag to this script.
+     # We now keep distinct sets of args, for a cleaner separation of concerns.
+
++    # add local rank for cpu-dist
++    sys.argv.append("--local_rank")
++    sys.argv.append(str(os.environ.get("PMI_RANK", -1)))
++
++    # ccl specific environment variables
++    if "ccl" in sys.argv:
```
Comment (hshen14) [inline intel_dist_training.md:282]: do we need to keep all ccl related stuff (e.g., import, and sys.argv) under if condition?

File: intel_dist_training.md [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+done
+export MASTER_ADDR=$master_addr
+
+source /home/ec2-user/anaconda3/envs/transformer/lib/python3.7/site-packages/torch_ccl-1.3.0+43f48a1-py3.7-linux-x86_64.egg/torch_ccl/env/setvars.sh
+
+export LD_LIBRARY_PATH=/home/ec2-user/anaconda3/envs/transformer/lib/python3.7/site-packages/torch_ccl-1.3.0+43f48a1-py3.7-linux-x86_64.egg/:$LD_LIBRARY_PATH
+export LD_PRELOAD="${CONDA_PREFIX}/lib/libtcmalloc.so:${CONDA_PREFIX}/lib/libiomp5.so"
+
+export CCL_WORKER_COUNT=4
+export CCL_WORKER_AFFINITY="0,1,2,3,32,33,34,34,35"
```
Comment (kding1) [inline intel_dist_training.md:246]: 34 is repeated.

File: intel_dist_training.md [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+<kbd>
+<img src="assets/36_intel_dist_training/04_four_nodes.png">
+</kbd>
+
+Training completes in __2 minutes and 36 seconds__, a __3x__ speedup.
+ 
+One last thing. Changing ```--task_name``` to ```qqp```, I also ran the Quora Question Pairs GLUE task, which is based on a much larger [dataset](https://quoradata.quora.com/First-Quora-Dataset-Release-Question-Pairs) (over 400,000 training samples). The fine-tuning times were:
+
+* Single-node: 11 hours 22 minutes,
+* 2 nodes: 7 hours and 52 minutes (1.45x),
```
Comment (kding1) [inline intel_dist_training.md:324]: this number seems slightly off from scaling efficiency perspective. I am getting a new measurement for your reference, should be ready tomorrow morning. it's trending to be 6 hour 38 min.

File: intel_dist_training.md (hunk 4 above)
Comment (juliensimon) [reply intel_dist_training.md:282]: Yes, we may want to use another backend, such as 'mpi'

File: intel_dist_training.md (hunk 5 above)
Comment (juliensimon) [reply intel_dist_training.md:246]: Geez.... Fixed.

File: intel_dist_training.md (hunk 6 above)
Comment (juliensimon) [reply intel_dist_training.md:324]: yes, I ran it again for a little while and it converged to a similar value.

File: intel_dist_training.md (hunk 1 above)
Comment (juliensimon) [reply intel_dist_training.md:1]: yes, I'll fix it.

File: intel_dist_training.md (hunk 3 above)
Comment (juliensimon) [reply intel_dist_training.md]: we'll keep working on this one, so I'll leave it for now

**Comment by juliensimon** [comment]
Thanks for the comments. I also added _blog_yml. I believe this is good to go.

File: _blog.yml [hunk 7]
```diff
@@ -215,4 +215,8 @@
 
 - local: bert-cpu-scaling-part-2
   title: "Scaling up BERT-like model Inference on modern CPU - Part 2"
-  date: November 4th, 2021
\ No newline at end of file
+  date: November 4th, 2021
+
+- local: accelerating-pytorch-distributed-fine-tuning
```
Comment (julien-c) [inline _blog.yml:220]: pet peeve of mine, but any way to do a shorter url?

like `accelerating-pytorch` or `faster-pytorch-with-intel` or... ?

Let's say less than 25 chars as a good guideline 😅

**Review by julien-c — CHANGES_REQUESTED** [review]

File: _blog.yml (hunk 7 above)
Comment (juliensimon) [reply _blog.yml]: no problem, renamed it to accelerating-pytorch.

**Review by julien-c — APPROVED** [review]

**Pull request merged**